### DISCUSSION TITLE
part-ii/2011-25/

### DISCUSSION THREAD

**Discussion by tripos-comments** [tripos-education/maths-tripos-questions#354] [Question comments]
Part II 2011 Coding and Cryptography: Paper 1, Section I, G 

https://questions.tripos.org/part-ii/2011-25/

**Comment by walrus1729** [comment]
<details>
<summary>Solution</summary>

Each question should halve the number of possibilities for $n$ (or as close as possible to halve, if we had an odd number of possibilities), so we define the questions we ask inductively as "Is $n \leq \frac{a+b-1}{2} + 1$?" where we have narrowed $n$ down to the interval $[a,b]$ (so we start with $a = 1, b = 10^6$ and our first question is "Is $a \leq 500 001$?"). This works because $2^{20} > 10^6$ (in fact $2^{20} = 1 048 576$).

The answer to the second part is no. There are $2^{20}$ possible sets of answers, so if we have more than $2^{20}$ possible orderings, then there must be two orderings that give the same set of answers and thus make it impossible to know the correct ordering for sure. But the number of possible orderings is $15!$, which is much larger than $2^{20}$ (this can be seen by the bound $15! \geq 2 \cdot 10 \cdot 11 \cdot 12 \cdot 13 \cdot 14 \cdot 15 \geq 2 \cdot 10^6 > 1 048 576 = 2^{20}$, or in various other ways).

Each question asked halves the number of possibilities (approximately, since we can have an odd number of possibilities), so the number of questions we need to ask is roughly $\log(n!)$ (where the logarithm is to the base $2$). For $n$ large, we can use Stirling's approximation 

$$
n! \approx \sqrt{2\pi n} (\frac{n}{e})^n
$$

So, after using some standard properties of the logarithm, the number of questions needed is approximately

$$
(n + \frac{1}{2})\log{n} - n\log{e} + \frac{1}{2}(\log{\pi} + 1)
$$

where all the logarithms are to the base $2$. (It is probably sufficient to only use the approximation $n! \approx (\frac{n}{e})^n$, which gives the estimate $n(\log{n} - \log{e})$.)

</details>
